### DISCUSSION TITLE
Route Groups: Organize routes without affecting the URL structure

### DISCUSSION THREAD

**Discussion by widgeter** [withastro/roadmap#480] [Proposal]
### Body

# Summary

Organize routes without affecting the URL structure using **Route Group**. This can be created by wrapping a folder’s name in parenthesis: _(folderName)_

# Background & Motivation

To better organize the routes in the `page` folder without the need to include subfolders that can affect the urls, something that could help is the use of **Route Groups**.

This is a concept that was introduced in **Next.js v13**. It is well documented at this url:

https://beta.nextjs.org/docs/routing/defining-routes#route-groups

**Astro** is a great framework and in the use of several projects we have noticed that when the number of pages increases, it becomes a huge list of resources without semantic order. With this new feature we could separate into folders (without affecting the urls) pages that belong to different modules or goals.

# Goals

- Organize routes without affecting the URL structure

# Example

```
/
├── src/
│  └── pages/
│       ├── (marketing)/
│       │   ├── about.astro
│       │   └── blog/
│       │       └── index.astro
│       ├── (shop)/
│       │   └── account.astro
│       ├── (legal)/
│       │   ├── terms.astro
│       │   └── privacy.astro
│       ├── index.astro
│       ├── 404.astro
│       ├-- rss.xml.ts
│       └── ...
└── ...
```

**Comment by pijusz** [comment]
Would love this enchancement in Astro!

**Comment by NuroDev** [comment]
By no means a priority but I really hope this gets added as a nice QoL improvement.

It makes organising a "full-stack site" (Marketing, dashboard, etc) project that bit easier

**Comment by cprcrack** [comment]
This would be great, any chance to reconsider it in 2025? :)

**Comment by florian-lefebvre** [comment]
FYi you can build this kind of custom abstraction in userland using https://inox-tools.fryuni.dev/custom-routing

**Reply by daimalou** [reply]
Thank you! I’ve solved. https://github.com/Fryuni/inox-tools/issues/224

**Comment by volkandkaya** [comment]
Came here after AI hallucinated that the feature exists. It is a pretty useful feature especially for larger sites.

**Reply by cnrad** [reply]
I'm here for the exact same reason haha. +1

**Comment by TSinChen** [comment]
Currently migrating a site to Astro from a legacy stack, and ran into this exact gap.

A few thoughts on whether this is still worth pursuing:

**`inox-tools/custom-routing` feels like a different tradeoff.** It's a great escape hatch when you need fully custom routing. But for the routes you want to group, having to enumerate each path-to-file mapping is a DX hit, and loses some of the simplicity of file-based routing. The point of `(folder)` is that it keeps the existing routing model while still letting you organize routes by feature or domain.

**Some concrete cases.** Pages like `/policy`, `/terms`, and `/trademark` are naturally related and would sit cleanly under `(legal)/`. Or routes could be grouped by feature like `(auth)/`, `(dashboard)/`, `(marketing)/`, each containing its own pages, `_components`, and other private resources. The current `src/pages/` structure doesn't really provide a natural place for that kind of organization.
